pinot-core/src/main/java/com/linkedin/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -367,20 +380,95 @@ private void setGroupByResults(@Nonnull BrokerResponseNative brokerResponseNativ
       }
       finalResultMaps[i] = finalResultMap;
     }
+    //If HAVING clause is set, we further filter the group by results based on the HAVING predicate
+    if (havingFilterQuery != null) {
+      HavingClauseComparisonTree havingClauseComparisonTree =
+          HavingClauseComparisonTree.buildHavingClauseComparisonTree(havingFilterQuery, havingFilterQueryMap);
+
+      //Applying close policy
+      //We just keep those groups (from different aggregation functions) that are exist in the result set of all aggregation functions.
+      //In other words, we just keep intersection of groups of different aggregation functions.
+      for (int i = 0; i < numAggregationFunctions; i++) {
+        Iterator<Map.Entry<String, Comparable>> finalMapIterator = finalResultMaps[i].entrySet().iterator();
```

Comment:
You should iterate on the keySet instead of the entrySet if you're not looking at the values.

It might be clearer to rewrite this to use `retainAll` to do set intersection instead of removing elements manually.